src/main/java/com/ncc/aif/AIFProgressMonitor.java

```diff
@@ -69,40 +85,68 @@ public void done() {
     @Override
     public void subTask(String label) {
         log("Validating " + label);
-        final String[] parts = label.split(" ");
-        final int shapeNum = Integer.parseUnsignedInt(parts[1].replaceFirst(":", ""));
-        final String shapeName = parts[2];
-        shapeNameMap.put(shapeNum, shapeName);
-        shapeTimeMap.put(shapeNum, System.currentTimeMillis());
+        shapeList.add(new ImmutablePair<>(label.split(" ")[2], System.currentTimeMillis()));
     }
 
     @Override
     public void worked(int amount) {
         final Date endTime = new Date();
         shapeNum += amount;
-        log("Worked " + shapeNum + " / " + numShapes);
-        final String shapeName = shapeNameMap.get(shapeNum);
-        final long startTimeMs = shapeTimeMap.get(shapeNum);
+        log("Completed shape " + shapeNum + " / " + numShapes);
+        final ImmutablePair<String, Long> pair = shapeList.remove();
+        final String shapeName = pair.getKey();
+        final long startTimeMs = pair.getValue();
         final long duration = endTime.getTime() - startTimeMs;
 
         if (duration > LOGGING_THRESHOLD) {
             log("  Duration: " + duration + "ms");
             log("  Timestamp: " + FORMAT.format(endTime));
         }
         // Write a row of the file.
+        if (out != null) {
+            try {
+                final String row = shapeNum + "\t" +
+                        shapeName + "\t" +
+                        FORMAT.format(new Date(startTimeMs)) + "\t" +
+                        FORMAT.format(endTime) + "\t" +
+                        duration + "\n";
+                out.write(row);
+                out.flush();
+                if (shapeNum >= numShapes) {
+                    out.close();
+                }
+            } catch (IOException ioe) {
+                logger.error("Could not write to progress monitor.");
+            }
+        }
+    }
+
+    // Unit test that simulates a validator invoking both variants of the AIFProgressMonitor
```

Comment:
Unit test should be in the src/test directory with the @Test directive so that it executes during a mvn test.

Reply:
I didn't get the impression that @ecurley-nextcentury was requesting a JUnit test to be added to the surefire test suite-- just a simple main program that shows it working.  We don't have JUnit tests for the StatsCollector (profiler), fail-fast, or other validator features.  For that matter, we don't have a unit test for the command-line validator.

Do you think a JUnit test for this class (or others) is called for?  And if so, could that be a separate ticket?

Reply:
I'll leave it up to @ecurley-nextcentury to clarify what he is asking for.  However, if the comment calls this a unit test, then it should be in the src/test directory. So, pending Eddie's clarification, either update the reference to it being a unit test, or move it and add the Test directive.

Reply:
Removed the reference to unit test.  If directed, I will move the test to the src/test directory and call it a unit test.

Reply:
After discussion, we decided to keep the main program test in the AIFProgressMonitor, but write a [new Jira ticket](https://nextcentury.atlassian.net/browse/AIDA-828) for JUnit tests for secondary validator functionality like the progress monitors and fail-fast.